### PR TITLE
Logging - Add per-layer grad norm logs

### PR DESCRIPTION
**Pull request by yzygitzh**
Adds per-layer grad norm logs for model training diagnosis.

Leverages `_ParamAndGradBuffer` from DDP to get gradients for all named parameters. Removes grad norm reducing in PP dimension to get per-layer results.

Currently only supports DistributedDataParallel. Use `--log-grad-norm-per-layer` to enable. For extra weights that don't belong to any layer, use `--log-grad-norm-per-layer-extra-patterns` to specify their naming patterns, e.g., `--log-grad-norm-per-layer-extra-patterns embedding output_layer final_layernorm`.

### CODE REVIEW

File: megatron/core/optimizer/optimizer.py [hunk 1]
```diff
@@ -192,6 +248,39 @@ def get_grad_norm(self):
...
+                grads_for_norm, grad_stats_parallel_group=parallel_state.get_tensor_model_parallel_group()
+            )
+            if isinstance(global_layer_id, str):
+                for idx, pattern in enumerate(extra_patterns):
+                    if pattern in global_layer_id:
+                        total_norm_per_layer[args.num_layers + idx] = total_norm
+                        break
+            else:
+                total_norm_per_layer[global_layer_id] = total_norm
+        # Get grad norms of all layers by reducing across PP dimension
```
Comment (yzygitzh) [inline megatron/core/optimizer/optimizer.py:280]: This is due to wandb is only trigger in the last rank so it needs to get the results from all PP ranks

File: megatron/core/optimizer/optimizer.py
Comment (yzygitzh) [inline megatron/core/optimizer/optimizer.py:1]: Add test

File: megatron/training/arguments.py [hunk 2]
```diff
@@ -967,6 +967,10 @@ def validate_args(args, defaults={}):
             + f"The supported position embedding types are rope and none."
         )
 
+    # Logging.
+    if args.log_grad_norm_per_layer:
+        assert not args.use_torch_fsdp2 and not args.use_custom_fsdp, "Per-layer grad norm logging only supports DistributedDataParallel."
```
Comment (cp5555) [inline megatron/training/arguments.py]: I think it means that `Per-layer grad norm logging doesn't support FSDP.` based on your code.

Comment (cp5555) [reply megatron/training/arguments.py]: Also, for logging, should we use assert to stop? I am not sure about it.

Comment (cp5555) [reply megatron/training/arguments.py]: If you only support DDP, why not check DDP directly?

File: megatron/core/optimizer/optimizer.py [hunk 3]
```diff
@@ -153,6 +156,59 @@ def get_main_grads_for_grad_norm(self) -> List[torch.Tensor]:
...
+                    if 'layers' in name:
+                        name_segments = name.split('.')
+                        local_layer_id = None
+                        for idx, param_name_segment in enumerate(name_segments):
+                            if param_name_segment == 'layers':
+                                local_layer_id = int(name_segments[idx + 1])
+                                break
+                        global_layer_id = global_layer_offset + local_layer_id
+                    else:
+                        global_layer_id = name
```
Comment (cp5555) [inline megatron/core/optimizer/optimizer.py:183]: Should we add global_layer information here? Otherwise, how to differentiate during distributed training? Does  this support TP, EP, etc?

File: megatron/core/optimizer/optimizer.py [hunk 4]
```diff
@@ -153,6 +156,59 @@ def get_main_grads_for_grad_norm(self) -> List[torch.Tensor]:
...
+                            if param in buffer.param_to_bucket:
+                                bucket = buffer.param_to_bucket[param]
+                                grad_data = bucket.grad_data
+                                param_start_index, param_end_index, _ = buffer.param_index_map[param]
+                                grad = grad_data[param_start_index - bucket.offset : param_end_index - bucket.offset]
+                                self.grads_for_norm_per_layer[global_layer_id].append(grad)
+                                grad_found = True
+                                break
+                        if grad_found:
+                            break
```
Comment (cp5555) [inline megatron/core/optimizer/optimizer.py:218]: There is only one grad for one global_layer_id?

Comment (cp5555) [reply megatron/core/optimizer/optimizer.py:218]: Also, should we keep updated for the grad?

File: megatron/core/optimizer/optimizer.py [hunk 5]
```diff
@@ -192,6 +248,39 @@ def get_grad_norm(self):
...
+        total_norm_per_layer_cuda = torch.tensor(total_norm_per_layer, dtype=torch.float, device='cuda')
+        torch.distributed.all_reduce(
+            total_norm_per_layer_cuda, op=torch.distributed.ReduceOp.SUM, group=parallel_state.get_pipeline_model_parallel_group()
+        )
+        total_norm_dict = {}
+        for idx in range(args.num_layers):
+            total_norm_dict[f'layer-{idx}'] = total_norm_per_layer_cuda[idx].item()
+        for idx, pattern in enumerate(extra_patterns):
+            total_norm_dict[pattern] = total_norm_per_layer_cuda[args.num_layers + idx].item()
+        return total_norm_dict
```
Comment (cp5555) [inline megatron/core/optimizer/optimizer.py:290]: Not fully understand this. Why is it much more complicated than `get_grad_norm()`

File: megatron/core/optimizer/optimizer.py (hunk 3 above)
Comment (yzygitzh) [reply megatron/core/optimizer/optimizer.py:183]: There are two types of parameter name:

- Contains local layer ID, like `*.layer.0.*`. In this case, we can get the global layer ID in the whole model.
- No local layer ID, like `embedding`. In this case its name is unique in the whole model and we leverage it.

Such naming rule is not related to how model is partitioned.

File: megatron/core/optimizer/optimizer.py (hunk 4 above)
Comment (yzygitzh) [reply megatron/core/optimizer/optimizer.py:218]: There are multiple parameters for one global_layer_id, and there is only one grad for one parameter. This for loop finds current parameter's grad and append it to the grad list of current global_layer_id.

File: megatron/core/optimizer/optimizer.py (hunk 5 above)
Comment (yzygitzh) [reply megatron/core/optimizer/optimizer.py:290]: They're different in two ways:
- `get_grad_norm` passes the grad list of the whole model to `get_grad_norm_fp32` and get its norm, but`get_grad_norm_per_layer` need to pass the grad list from different layers one by one to get separate norms.
- Different PP rank only have norms for layers in local PP stage, but wandb logging happened in the last global rank, so we need to do one accumulation to let last rank get norms of all layers.

File: megatron/training/arguments.py (hunk 2 above)
Comment (yzygitzh) [reply megatron/training/arguments.py]: Changed to detect DDP and print warning when condition is not met. DDP condition is `args.use_distributed_optimizer and not args.use_custom_fsdp` according to existing argument definition.

File: megatron/core/optimizer/optimizer.py (hunk 4 above)
Comment (yzygitzh) [reply megatron/core/optimizer/optimizer.py:218]: For gradient update, underlying grad buffer data is already updated by DDP optimizer when we this function is called.

It's also verified by the unit test added. Original grad norm result and new grad norm per layer result are numerically consistent in two training steps.

File: megatron/training/training.py [hunk 6]
```diff
@@ -1476,6 +1479,15 @@ def training_log(loss_dict, total_loss_dict, learning_rate, decoupled_learning_r
             if wandb_writer:
                 wandb_writer.log({'grad-norm': grad_norm}, iteration)
                 wandb_writer.log({'grad-norm-with-scale': grad_norm * loss_scale}, iteration)
+        if grad_norm_per_layer is not None:
+            for key in grad_norm_per_layer:
+                grad_norm_value = grad_norm_per_layer[key]
+                writer.add_scalar(f'grad-norm-{key}', grad_norm_value, iteration)
```
Comment (cp5555) [inline megatron/training/training.py:1485]: Why do you include grad-norm-with-scale?

File: megatron/training/arguments.py [hunk 7]
```diff
@@ -967,6 +967,10 @@ def validate_args(args, defaults={}):
             + f"The supported position embedding types are rope and none."
         )
 
+    # Logging.
+    if args.log_grad_norm_per_layer and not (args.use_distributed_optimizer and not args.use_custom_fsdp):
```
Comment (cp5555) [inline megatron/training/arguments.py:971]: Should we reset `args.log_grad_norm_per_layer`

File: megatron/training/training.py [hunk 8]
```diff
@@ -1972,6 +1984,10 @@ def get_e2e_base_metrics():
         torch.distributed.barrier()
         print_rank_0(f">>> Weight hashes match after {iteration} iterations...")
 
+    # Prepare global layer offset for per-layer grad norm.
+    if args.log_grad_norm_per_layer and (args.use_distributed_optimizer and not args.use_custom_fsdp):
```
Comment (cp5555) [inline megatron/training/training.py]: Should we check (args.use_distributed_optimizer and not args.use_custom_fsdp)? I would like to reset `args.log_grad_norm_per_layer` in the beginning.

File: megatron/training/training.py [hunk 9]
```diff
@@ -2083,12 +2099,17 @@ def get_e2e_base_metrics():
                 decoupled_learning_rate = param_group['lr']
             else:
                 learning_rate = param_group['lr']
+        grad_norm_per_layer = None
+        if args.log_grad_norm_per_layer and (args.use_distributed_optimizer and not args.use_custom_fsdp):
```
Comment (cp5555) [inline megatron/training/training.py]: Should we reset `args.log_grad_norm_per_layer` directly?

File: megatron/training/training.py (hunk 6 above)
Comment (yzygitzh) [reply megatron/training/training.py:1485]: To keep aligned with previous whole-model grad norm logging

**Review by cp5555 — APPROVED** [review]

File: megatron/training/training.py (hunk 9 above)
Comment (yzygitzh) [reply megatron/training/training.py]: Fixed

File: megatron/training/training.py (hunk 8 above)
Comment (yzygitzh) [reply megatron/training/training.py]: Fixed

File: megatron/training/arguments.py (hunk 7 above)
Comment (yzygitzh) [reply megatron/training/arguments.py:971]: Fixed

File: megatron/core/optimizer/optimizer.py [hunk 10]
```diff
@@ -153,6 +153,72 @@ def get_main_grads_for_grad_norm(self) -> List[torch.Tensor]:
 
         return grads_for_norm
 
+    def get_main_grads_for_grad_norm_per_layer(self, global_layer_offset) -> Dict[Union[int, str], List[torch.Tensor]]:
+        """
+        Get per-layer main_grads that should be taken into account to compute the grad norm.
+        Currently only support DistributedDataParallel (DDP).
+        """
+        if not hasattr(self, 'grads_for_norm_per_layer'):
```
Comment (yukirora) [inline megatron/core/optimizer/optimizer.py:161]: why need this line?

Comment (yzygitzh) [reply megatron/core/optimizer/optimizer.py:161]: The calculation for `grads_for_norm_per_layer` is expensive so we would like to cache it. It's possible because `grads_for_norm_per_layer` is a group of tensor views which will not change in the future.

**Review by yukirora — APPROVED** [review]
